src/Control/Lens/Iso.hs

```diff
@@ -397,6 +398,32 @@ anon a p = iso (fromMaybe a) go where
        | otherwise = Just b
 {-# INLINE anon #-}
 
+-- | Improper 'non' - this is a version of 'non' suitable for caching.
+-- Use 'inon' instead of 'non' to manage a map of values where you don't
+-- want to lose information about which keys were ever present in the map.
+--
+-- The expression @'non' a@ maps @Nothing@ to @a@ and @a@ back to @Nothing@,
+-- so it is a proper 'Iso'.  Instead, @'inon' a@ still maps @Nothing@
+-- to @a@, but in the reverse direction it maps @a@ back to @Just a@.
+-- If you have a map of values such as @Map UserId UserProfile@ which
+-- you are accessing using 'non' with a lens like @at uid . non def
+-- . somefield@, when you set that field to a value that happens to
+-- restore the condition @UserProfile == def@, the record gets deleted
+-- from the map.  You have now lost the information that that user was
+-- ever present in the map:
+--
+-- >>> fromList [("user1", "!!!")] & (at "user1" . non def) .~ ""
+-- fromlist []
+--
+-- Using 'inon' instead of 'non' means that key will not be deleted
+-- from the map:
+--
+-- >>> fromList [("user1", "!!!")] & (at "user1" . inon def) .~ ""
```

Comment:
Don't

```haskell
fromList [("user1", "!!!")] & at "user1" ?~ ""
```

has same effect?

And for caching, why the removal of defautl value would be bad. I don't understand that either.

Reply:
I'll answer the second question.  The scenario is that your cache is a Map, and the set of keys is an important piece of information, you don't want to forget that a key was once an element of the map.  Picture the user list of a web app, you don't want the user to disappear if they reset their user profile to the default.

Reply:
As for the first question, yes that has the same effect, but in my use case I am constructing the lens with non and then composing further lenses to look inside the value.  This lens gets passed to a function, so we no longer have access to the lens that looks at the map itself.

Reply:
like this:

```
>>> (fromList [("user1", "axc")]) & (at "user1" . non "abc" . ix 1) .~ 'b'`
fromList []
```